### PR TITLE
adding vagrant-aws support

### PR DESCRIPTION
**Pull request by blakefrantz**
add vagrant-aws support per issue #1144

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: Vagrantfile [hunk 1]
```diff
@@ -1,3 +1,58 @@
+
+
```
Comment (marpaia) [inline Vagrantfile]: Can you remove this and all other superfluous whitespace?

File: Vagrantfile [hunk 2]
```diff
@@ -1,3 +1,58 @@
...
+targets = {
+
+  #
+  # the following targets are used with local vagrant providers (i.e virtualbox). 
+  #
+
+  "centos6.5" => { "box" => "chef/centos-6.5" },
```
Comment (marpaia) [inline Vagrantfile]: can you expand these to look like:

```
TARGETS = {
  "centos6.5" => {
    "box" => "chef/centos-6.5",
  },
  # etc
}
```

File: Vagrantfile [hunk 3]
```diff
@@ -1,3 +1,58 @@
...
+  "centos7"   => { "box" => "chef/centos-7.0" },
+  "ubuntu14"  => { "box" => "ubuntu/trusty64" },    
+  "ubuntu12"  => { "box" => "ubuntu/precise64" },
+  "freebsd10" => { "box" => "chef/freebsd-10.0" },
+
+  #
```
Comment (marpaia) [inline Vagrantfile]: remove these comments, they're unnecessary

File: Vagrantfile [hunk 4]
```diff
@@ -1,3 +1,58 @@
...
+    "box" => "andytson/aws-dummy",
+    "regions" => { "us-east-1" => "ami-12663b7a", "us-west-1" => "ami-a540a5e1" },
+    "username" => "ec2-user"
+  },
+  "aws-rhel6.5" => {
+    "box" => "andytson/aws-dummy", 
+    "regions" => { "us-east-1" => "ami-1643ff7e", "us-west-1" => "ami-2b171d6e" },
+    "username" => "ec2-user"
+  },
+  "aws-centos7" => {
```
Comment (marpaia) [inline Vagrantfile]: what was the motivation behind duplicating all of these operating systems? we already have vagrant boxes for them.

File: Vagrantfile [hunk 5]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
-   %w{centos7   chef/centos-7.0},
-   %w{ubuntu14  ubuntu/trusty64},
-   %w{ubuntu12  ubuntu/precise64},
-   %w{freebsd10 chef/freebsd-10.0}
- ].each do |machine|
-  (name, box) = machine
+
+  config.vm.provider :aws do |aws, override|
+
+    # Required. Credentials for aws api. vagrant-aws will error if these are unset.
```
Comment (marpaia) [inline Vagrantfile]: will this work if you try to vagrant up a !AWS box?

File: Vagrantfile [hunk 6]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+    aws.access_key_id = ENV['AWS_ACCESS_KEY_ID']
+    aws.secret_access_key = ENV['AWS_SECRET_ACCESS_KEY']
```
Comment (marpaia) [inline Vagrantfile]: will this work if you try to vagrant up a !AWS box?

File: Vagrantfile [hunk 7]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
-   %w{centos6.5 chef/centos-6.5},
-   %w{centos7   chef/centos-7.0},
-   %w{ubuntu14  ubuntu/trusty64},
-   %w{ubuntu12  ubuntu/precise64},
-   %w{freebsd10 chef/freebsd-10.0}
- ].each do |machine|
-  (name, box) = machine
+
+  config.vm.provider :aws do |aws, override|
+
```
Comment (marpaia) [inline Vagrantfile]: rm whitespace

File: Vagrantfile [hunk 8]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+    # Required. Credentials for aws api. vagrant-aws will error if these are unset.
+    aws.access_key_id = ENV['AWS_ACCESS_KEY_ID']
+    aws.secret_access_key = ENV['AWS_SECRET_ACCESS_KEY']
+
+    # Required. Name of aws keypair for launching and accessing the ec2 instance.
+    if [ ENV['AWS_KEYPAIR_NAME'] ]
+      aws.keypair_name = ENV['AWS_KEYPAIR_NAME']
+    end
+
+    # Path to private key for above keypair_name.
```
Comment (marpaia) [inline Vagrantfile]: rm comment, "AWS_SSH_PRIVATE_KEY_PATH" is pretty descriptive IMO

File: Vagrantfile [hunk 9]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
- ].each do |machine|
-  (name, box) = machine
+
+  config.vm.provider :aws do |aws, override|
+
+    # Required. Credentials for aws api. vagrant-aws will error if these are unset.
+    aws.access_key_id = ENV['AWS_ACCESS_KEY_ID']
+    aws.secret_access_key = ENV['AWS_SECRET_ACCESS_KEY']
+
+    # Required. Name of aws keypair for launching and accessing the ec2 instance.
```
Comment (marpaia) [inline Vagrantfile]: if this is required, why is it in an if like this, but not the rest of them?

File: Vagrantfile [hunk 10]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+    if [ ENV['AWS_KEYPAIR_NAME'] ]
```
Comment (marpaia) [inline Vagrantfile]: if this is required, why is it in an if like this, but not the rest of them?

File: Vagrantfile [hunk 11]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+    # Set this to the desired aws instance.
+    if ENV['AWS_INSTANCE_TYPE']
+      aws.instance_type = ENV['AWS_INSTANCE_TYPE']
+    else
+      aws.instance_type = "m3.large"
+    end
+
+    # hack: make aws.region accessible in targets.each do loop.
+    targets["active_region"] = aws.region
+
```
Comment (marpaia) [inline Vagrantfile]: rm whitespace

File: Vagrantfile [hunk 12]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+        aws.ami = target['regions'][targets["active_region"]]
+
+        #
+        # user_data is a shell script that is executed by cloud-init early in
+        # the boot process. this script exempts the default ssh user from 
+        # requiretty directive typically set in /etc/sudoers. if this is not 
+        # done, rsync breaks.
+        #
+
+        aws.user_data = "#!/bin/bash\necho 'Defaults:" + target['username'] + " !requiretty' > /etc/sudoers.d/999-vagrant-cloud-init-requiretty && chmod 440 /etc/sudoers.d/999-vagrant-cloud-init-requiretty "
```
Comment (marpaia) [inline Vagrantfile]: we definitely need to do something cleaner here. maybe a script in tools that you read in?

File: Vagrantfile [hunk 13]
```diff
@@ -33,5 +146,5 @@ Vagrant.configure("2") do |config|
       end
     end
 
-  end
+  end
```
Comment (marpaia) [inline Vagrantfile]: whitespace at the end of the line here?

File: Vagrantfile [hunk 14]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+        # user_data is a shell script that is executed by cloud-init early in
+        # the boot process. this script exempts the default ssh user from 
+        # requiretty directive typically set in /etc/sudoers. if this is not 
+        # done, rsync breaks.
+        #
+
+        aws.user_data = "#!/bin/bash\necho 'Defaults:" + target['username'] + " !requiretty' > /etc/sudoers.d/999-vagrant-cloud-init-requiretty && chmod 440 /etc/sudoers.d/999-vagrant-cloud-init-requiretty "
+
+        # OSs have varying default ssh usernames (ec2-user, ubuntu, etc)
+        override.ssh.username = target['username']
```
Comment (marpaia) [inline Vagrantfile]: why is this necessary? all of the usernames that you define look like the defaults, no?

File: Vagrantfile [hunk 15]
```diff
@@ -7,18 +62,76 @@ Vagrant.configure("2") do |config|
...
+
+        # Define tags for the instance
```
Comment (marpaia) [inline Vagrantfile]: rm comment

**Comment by marpaia** [comment]
we're getting pretty intense with environment variables here. perhaps it's worth making a "vagrant.conf" or something?

**Comment by theopolis** [comment]
I love this as a solution for Amazon Linux and RHEL development/testing. I'd like to hook our Jenkins up to build RHEL/Amazon packages using EC2 too. Supporting and scaling our own RHEL licensing is not going to work, we already tried and cannot automatically scale without hundreds to thousands of dollars worth of licensing. 

Duplicating the existing CentOS/Linux/FreeBSD targets in AWS might not present a good development/support UX. My fear with the duplicate targets is us not testing on them and introducing testing/deployment confounds from our existing targets being updated in the chef repos and not in the AMIs.

File: Vagrantfile (hunk 4 above)
Comment (blakefrantz) [inline Vagrantfile]: having the aws-based targets allows devs to spin up multiple vms w/o consuming local resources on their laptop. i have no problem removing the dups if folks prefer that.

File: Vagrantfile (hunk 10 above)
Comment (blakefrantz) [inline Vagrantfile]: nice catch. for stock amis, the keypair is required or you end up with a vm you cant login to. the vagrant-aws plugin issues a warning when the keypair is not set but does not fail hard. its possible that someone could build an ami that has preset creds that obviate the need for the keypair. given that a clear warning is issued by vagrant-aws, i removed the "required" label and kept the if block. let me know if there's a better route.

File: Vagrantfile (hunk 1 above)
Comment (blakefrantz) [inline Vagrantfile]: yup. done.

File: Vagrantfile (hunk 2 above)
Comment (blakefrantz) [inline Vagrantfile]: yup. done.

File: Vagrantfile (hunk 14 above)
Comment (blakefrantz) [inline Vagrantfile]: it's my understanding that typical vagrant targets are configured to allow vagrant to login as the user "vagrant". that "vagrant" user is allowed to sudo w/o a password.  these aws targets are not purpose built vagrant targets and do not have a "vagrant" user. they are sock images. however, their default usernames are configured to allow sudo w/o a password. our options are to either 1) build amis that have a "vagrant" user or to 2) tell the vagrant-aws plugin to use a different username to ssh in. 

pros/cons:

* option 1 pros 
 * platform specific usernames in the vagrant file go away
 * the aws.user_data line can go away
* option 1 cons
 * we need to build amis for each os we want an aws-based target for.
 * we need to replicate those ami to all aws regions that we want this to work in.
 * the aws account hosting the amis will incur small ebs fees.
* option 2 pros
 * we dont have to create or replicate amis
 * expanding into other aws regions is easier, just lookup the ami-id and put it in the targets hash
 * no ebs fees
* option 2 cons
 * the aws.user_data work around is required for rsync to work
 * the vagrant file has usernames in it.

there may be other pros/cons/options. what's your take?

**Comment by blakefrantz** [comment]
@theopolis @marpaia - see my comments above about why the additional targets. if those comments are not compelling let me know and i will remove all aws targets other than rhel/amazon. i'm pretty indifferent.

File: Vagrantfile (hunk 11 above)
Comment (blakefrantz) [inline Vagrantfile]: all whitespace has been mowed.

File: Vagrantfile (hunk 8 above)
Comment (blakefrantz) [inline Vagrantfile]: removed

File: Vagrantfile (hunk 6 above)
Comment (blakefrantz) [inline Vagrantfile]: yes, i've tested this with the local centos7, centos6.5, ubuntu12, ubuntu14, and freebsd10 targets.

**Comment by blakefrantz** [comment]
@marpaia regarding the envars - i started with the envars to avoid stuffing the access keys in the vagrantfile. those two access-key related envars are also usable by the aws-cli. we can drop the other envars and require users to manually update the vagrant file with their prefs. that might clean things up a bit. what's your take?

**Comment by blakefrantz** [comment]
@theopolis @marpaia  new [commit](https://github.com/blakefrantz/osquery/commit/72a3369e2b2a2b9108bbc893d44d2109cc5eb85d) submitted that resolves everything but the following open items:

1. remove/keep envars for non access-key items. i'm indifferent.
2. remove/keep aws-based targets that already have local targets. i'm indifferent.
3. keep userdata as it is or update vagrant file to read in a script and sub in the applicable username.
4. build/replicate custom amis or not

let me know what you think.

**Comment by marpaia** [comment]
1. I don't think forcing people to update the Vagrantfile is acceptable. It should either be environment variables or a configuration file separate from the Vagrantfile. I'd be OK with env vars if we also updated https://github.com/facebook/osquery/blob/master/docs/wiki/development/building.md

2. I think we should definitely remove them, personally. I can't find your "comments above about why the additional targets", maybe my github foo is failing me :) do you mind reposting here?

3. This definitely shouldn't stay in there as is, 100%. I would also be ok with a multi-line string inline in the file (http://commandercoriander.net/blog/2014/11/09/a-multiline-string-cheatsheet-for-ruby/).

4. I'm not too sure what you mean by this one?

**Comment by blakefrantz** [comment]
1. I'm happy to update the docs/wiki. i'll add that to this pr.
2. i'll pull the dups. my [original motivation](https://github.com/facebook/osquery/pull/1145#discussion_r30415355) was to offload running multiple vms to aws vice local laptop. 
3. good call. i'll go that route.
4. whether we go with option 1 (build custom amis) or option 2 (not do that) from [this comment](https://github.com/facebook/osquery/pull/1145#discussion_r30419701). i lean toward not building them.

File: Vagrantfile [hunk 16]
```diff
@@ -7,18 +24,50 @@ Vagrant.configure("2") do |config|
     end
     v.memory = 4096
   end
- [
-   %w{centos6.5 chef/centos-6.5},
-   %w{centos7   chef/centos-7.0},
-   %w{ubuntu14  ubuntu/trusty64},
-   %w{ubuntu12  ubuntu/precise64},
-   %w{freebsd10 chef/freebsd-10.0}
```
Comment (theopolis) [inline Vagrantfile]: Do we plan to add these back?

Comment (blakefrantz) [inline Vagrantfile]: Oops! Readded.

**Comment by theopolis** [comment]
Hey @blakefrantz, do you mind rebasing this branch from origin/master and testing the RHEL6 on EC2 again? I've added a few changes so build using the non-standard GCC toolchain:
- https://github.com/facebook/osquery/blob/master/CMake/FindGlog.cmake#L19
- https://github.com/facebook/osquery/blob/master/CMakeLists.txt#L94

**Comment by blakefrantz** [comment]
@theopolis  i've rebased, make deps passes but make fails on the same crtbegin error. the 2nd subsequent make attempt ends up working. i'm not sure why yet :)

**Comment by theopolis** [comment]
Ok, let's try again with this branch on the RHEL6.5 EC2 node.

**Comment by blakefrantz** [comment]
same errors. however, if i run the following pre-make then everything works great on RHEL6.5 EC2 node.
```
export CFLAGS=--gcc-toolchain=/opt/rh/devtoolset-3/root/usr/
export CXXFLAGS=--gcc-toolchain=/opt/rh/devtoolset-3/root/usr/
```

**Comment by theopolis** [comment]
Darn, thought we had it. I finally set up my AWS API keys and what not, working from my WIP branch that includes your commits: https://github.com/theopolis/osquery/commits/vagrant-aws. I'll testing builds on RHEL6.5 right now, it'll take me a few hours to get all the deps built.

**Comment by theopolis** [comment]
Merging in #1162.
